### PR TITLE
ticdc/pipeline: puller table actor refactor

### PR DESCRIPTION
**Pull request by sdojjy**
### What problem does this PR solve? 
this is a part of table pipeline actor model refactor  https://github.com/pingcap/ticdc/issues/3881

### What is changed and how it works?
add actorNodeContext and implement the NodeContext interface, so puller can use this context to send message to actor system.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update key monitor metrics in both TiCDC document and official document

### Release note 

```release-note
 `None`.
```

### CODE REVIEW

File: cdc/processor/pipeline/system/system.go [hunk 1]
```diff
@@ -42,3 +45,21 @@ func (s *System) Start(ctx context.Context) error {
...
+func (s *System) Router() *actor.Router {
+	return s.tableActorRouter
+}
+
+func (s *System) System() *actor.System {
+	return s.tableActorSystem
+}
+
+// ActorID returns an ActorID correspond with tableID.
+func ActorID(changefeedID string, tableID model.TableID) actor.ID {
```
Comment (overvenus) [inline cdc/processor/pipeline/system/system.go:58]: What about hash collision? Could you generate ID from a global id generator? E.g., an increase-only atomic variable.

File: cdc/processor/pipeline/actor_node_context.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/pingcap/ticdc/pkg/actor/message"
+	"github.com/pingcap/ticdc/pkg/context"
+	"github.com/pingcap/ticdc/pkg/pipeline"
+	"go.uber.org/zap"
+)
+
+// send a tick message to actor if we get 32 pipeline messages
+const messagesPerTick = 32
+
+type actorNodeContext struct {
```
Comment (overvenus) [inline cdc/processor/pipeline/actor_node_context.go:31]: Could you add some comment about the context?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: cdc/processor/pipeline/actor_node_context.go (hunk 2 above)
Comment (sdojjy) [reply cdc/processor/pipeline/actor_node_context.go:34]: done

File: cdc/processor/pipeline/system/system.go (hunk 1 above)
Comment (sdojjy) [reply cdc/processor/pipeline/system/system.go]: done

File: cdc/processor/pipeline/actor_node_context.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/pingcap/ticdc/pkg/actor/message"
+	"github.com/pingcap/ticdc/pkg/context"
+	"github.com/pingcap/ticdc/pkg/pipeline"
+	"go.uber.org/zap"
+)
+
+// send a tick message to actor if we get 32 pipeline messages
+const messagesPerTick = 32
+
+// actorNodeContext implements the NodeContext interface, with this we do not need to change too much logic to implement the table actor
```
Comment (asddongmen) [inline cdc/processor/pipeline/actor_node_context.go:31]: It is better that to limit the length of one line comment or code no more than 80 -- 100 chars.

File: cdc/processor/pipeline/actor_node_context.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	outputCh             chan pipeline.Message
+	tableActorRouter     *actor.Router
+	tableActorID         actor.ID
+	changefeedVars       *context.ChangefeedVars
+	globalVars           *context.GlobalVars
+	tickMessageThreshold int32
+	noTickMessageCount   int32
+}
+
+func NewContext(stdCtx sdtContext.Context, tableActorRouter *actor.Router, tableActorID actor.ID, changefeedVars *context.ChangefeedVars, globalVars *context.GlobalVars) *actorNodeContext {
```
Comment (asddongmen) [inline cdc/processor/pipeline/actor_node_context.go:45]: ditto

File: cdc/processor/pipeline/system/system.go [hunk 5]
```diff
@@ -15,19 +15,28 @@ package system
...
 
+	"github.com/pingcap/ticdc/cdc/model"
 	"github.com/pingcap/ticdc/pkg/actor"
 )
 
 // System manages table pipeline global resource.
 type System struct {
 	tableActorSystem *actor.System
 	tableActorRouter *actor.Router
+	idLock           sync.Mutex
```
Comment (asddongmen) [inline cdc/processor/pipeline/system/system.go:29]: Could add some comments for these id related fields? Like watch this ids stand for.

File: cdc/processor/pipeline/actor_node_context.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	if err == nil {
+		return
+	}
+	log.Error("puller stopped", zap.Error(err))
+	_ = c.tableActorRouter.SendB(c, c.tableActorID, message.StopMessage())
+}
+
+// SendToNextNode send msg to the outputCh and notify the actor system,
+// to reduce the  actor message, only send tick message per threshold
+func (c *actorNodeContext) SendToNextNode(msg pipeline.Message) {
```
Comment (asddongmen) [inline cdc/processor/pipeline/actor_node_context.go:80]: Do you think it is better that we send a tick message forcedly when received a `resolvedTs` msg.

File: cdc/processor/pipeline/actor_node_context.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+// the Throw function handle error and stop the actor
+type actorNodeContext struct {
+	sdtContext.Context
+	outputCh             chan pipeline.Message
+	tableActorRouter     *actor.Router
+	tableActorID         actor.ID
+	changefeedVars       *context.ChangefeedVars
+	globalVars           *context.GlobalVars
+	tickMessageThreshold int32
+	noTickMessageCount   int32
```
Comment (asddongmen) [inline cdc/processor/pipeline/actor_node_context.go:42]: The field name is a little confusing.

File: cdc/processor/pipeline/actor_node_context.go (hunk 3 above)
Comment (sdojjy) [reply cdc/processor/pipeline/actor_node_context.go:31]: done

File: cdc/processor/pipeline/actor_node_context.go (hunk 4 above)
Comment (sdojjy) [reply cdc/processor/pipeline/actor_node_context.go:45]: done

File: cdc/processor/pipeline/system/system.go (hunk 5 above)
Comment (sdojjy) [reply cdc/processor/pipeline/system/system.go:29]: done

File: cdc/processor/pipeline/actor_node_context.go (hunk 6 above)
Comment (sdojjy) [reply cdc/processor/pipeline/actor_node_context.go:85]: I will do some tests for that after this feature is implemented

File: cdc/processor/pipeline/actor_node_context.go (hunk 7 above)
Comment (sdojjy) [reply cdc/processor/pipeline/actor_node_context.go]: added comment

**Review by asddongmen — APPROVED** [review]

**Review by overvenus — APPROVED** [review]

**Pull request merged**